runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java

```diff
@@ -676,7 +676,9 @@ private synchronized void updateTimers(TimerUpdate update) {
       Map<StructuralKey<?>, List<TimerData>> firedTimers;
       switch (domain) {
         case PROCESSING_TIME:
-          firedTimers = extractFiredTimers(firingTime, processingTimers);
+          firedTimers =
+              extractFiredTimers(
+                  INSTANT_ORDERING.min(firingTime, earliestHold.get()), processingTimers);
```

Comment:
I don't think this is right. We want to fire timers based on firing time, not based on the hold.

Reply:
done